### PR TITLE
build/ops: fix dmclock test-related build breakage by conditionalizing dmclock tests

### PR DESCRIPTION
**Pull request by smithfarm**
This is an alternative PR to https://github.com/ceph/ceph/pull/14997 (that PR builds only on the most cutting-edge openSUSE version - Tumbleweed . . . still trying to figure out exactly why)

http://tracker.ceph.com/issues/19987

### CODE REVIEW

**Comment by smithfarm** [comment]
Pushed `wip-findgtest-cmake` to ceph/ceph-ci.git

**Comment by smithfarm** [comment]
Relevant excerpt from log of failed build:

```
[  158s] CMake Error at src/dmclock/CMakeLists.txt:20 (find_package):
[  158s]   By not providing "Findgtest.cmake" in CMAKE_MODULE_PATH this project has
[  158s]   asked CMake to find a package configuration file provided by "gtest", but
[  158s]   CMake did not find one.
[  158s] 
[  158s]   Could not find a package configuration file provided by "gtest" with any of
[  158s]   the following names:
[  158s] 
[  158s]     gtestConfig.cmake
[  158s]     gtest-config.cmake
[  158s] 
[  158s]   Add the installation prefix of "gtest" to CMAKE_PREFIX_PATH or set
[  158s]   "gtest_DIR" to a directory containing one of the above files.  If "gtest"
[  158s]   provides a separate development package or SDK, be sure it has been
[  158s]   installed.
[  158s] 
[  158s] 
[  158s] -- Configuring incomplete, errors occurred!
```

**Comment by tchaikov** [comment]
@smithfarm once ceph-ci is happy, it's good to merge, i think.

**Comment by smithfarm** [comment]
@tchaikov Thanks, but SUSE builds are still not happy. Now the build error is:

```
[  175s] CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:148 (message):
[  175s]   Could NOT find GTest (missing: GTEST_LIBRARY GTEST_INCLUDE_DIR
[  175s]   GTEST_MAIN_LIBRARY)
[  175s] Call Stack (most recent call first):
[  175s]   /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:388 (_FPHSA_FAILURE_MESSAGE)
[  175s]   /usr/share/cmake/Modules/FindGTest.cmake:205 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
[  175s]   src/dmclock/CMakeLists.txt:20 (find_package)
```

File: src/dmclock/CMakeLists.txt [hunk 1]
```diff
@@ -17,7 +17,7 @@ if(K_WAY_HEAP)
 endif()
 
 if (NOT(TARGET gtest AND TARGET gtest_main))
-  find_package(gtest REQUIRED)
+  find_package(GTest REQUIRED)
```
Comment (tchaikov) [inline src/dmclock/CMakeLists.txt]: we do include GTest as a submodule. can we point `GTEST_ROOT` to the path in the source tree, so `FindGTest.cmake` can find it?

Comment (smithfarm) [inline src/dmclock/CMakeLists.txt]: OK, how?

**Comment by smithfarm** [comment]
OK, dropped that last commit.

**Comment by smithfarm** [comment]
@tchaikov Something like this?

```
set(GTEST_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/googletest)
```

**Comment by tchaikov** [comment]
@smithfarm something like that, but i am not 100% sure. because at configure time (when we run `cmake`) the gtest lib is not ready yet, so `FindGtest` could fail. that's a dilemma i was facing when trying to make boost an external project, which is built on demand. i need to give it a try locally next Monday.

**Comment by smithfarm** [comment]
Same error. I'll continue to poke at it.

**Comment by smithfarm** [comment]
I think the build failure is reproducible by passing -DWITH_TESTS=OFF to cmake.

**Comment by smithfarm** [comment]
This fixes the `-DWITH_TESTS=OFF` build.

**Comment by smithfarm** [comment]
Looks like we can drop https://github.com/ceph/ceph/blob/master/src/dmclock/cmake/modules/Findgtest.cmake

**Comment by smithfarm** [comment]
@ivancich @tchaikov Here is how we could conditionalize the build of the dmclock test binaries. (Note that WITH_TESTS is not appropriate here, because it guards the build of binaries included in the "ceph-test" subpackage.)

If we're going to build these dmclock test binaries separately, my view is they should be properly conditionalized. I understand this may mean including the conditional in the upstream dmClock repo as well.

**Comment by smithfarm** [comment]
note that this PR now consists of a single (completely reworked) commit

**Comment by ivancich** [comment]
My concern regarding this PR is that it conflicts with the PR for the second part of the dmclock integration. That PR, for example, currently does not integrate src/dmclock/CMakeLists.txt with ceph's build, so the changes there will be problematic. And that removes the need for any ceph conditionals to appear within the dmclock library, which seems inappropriate.

File: src/dmclock/CMakeLists.txt [hunk 2]
```diff
@@ -16,10 +16,12 @@ if(K_WAY_HEAP)
   endif()
 endif()
 
-if (NOT(TARGET gtest AND TARGET gtest_main))
-  find_package(gtest REQUIRED)
-  include_directories(${GTEST_INCLUDE_DIRS})
-endif()
+if(WITH_DMCLOCK_TESTS)
```
Comment (ivancich) [inline src/dmclock/CMakeLists.txt]: WITH_DMCLOCK_TESTS should not appear below src/dmclock. A better approach is the one used in https://github.com/ceph/ceph/pull/14997. In fact a good starting point might be the final two commits in that PR. The first is a git subtree pull of the latest dmclock master that significantly reorganizes dmclock's cmake process (see README.git-subtree to see how this is done). The second (commit 8215686) are corresponding changes in ceph's cmake process. This approach as a starting point will avoid conflicts and re-visiting this issue down the line.

**Comment by ivancich** [comment]
@smithfarm I'm currently testing a slight change to https://github.com/ceph/ceph/pull/14997, which I think will handle your case as well. I'm currently testing the build. The additions to src/CMakeLists.txt are below and you can also find the entire branch here: https://github.com/ivancich/ceph-fork/tree/wip-bring-in-dmclock-p2-leap . 

```
option(WITH_DMCLOCK_TESTS                                                                                                                                                             
  "enable the build of dmclock-tests and dmclock-data-struct tests binaries"                                                                                                          
  OFF)                                                                                                                                                                                
if(WITH_TESTS AND WITH_DMCLOCK_TESTS)                                                                                                                                                 
  # note: add_test is not being called, so dmclock tests aren't part                                                                                                                  
  # of ceph tests                                                                                                                                                                     
  add_subdirectory(dmclock/test)                                                                                                                                                      
  add_subdirectory(dmclock/support/test)                                                                                                                                              
endif(WITH_TESTS AND WITH_DMCLOCK_TESTS)  
```

**Comment by ivancich** [comment]
I've added a WITH_DMCLOCK_TESTS commit (https://github.com/ceph/ceph/pull/14997/commits/04071b618db3c38ca664851ffbc4676f69924f9d) to PR for part 2 of the dmclock integration (https://github.com/ceph/ceph/pull/14997). It also includes the changes you recommended regarding cmake's find_package. @smithfarm does that work for you?

**Comment by smithfarm** [comment]
@ivancich The latest #14997 fixes the problem for us; we're going with it - thanks!

**Pull request closed without merging**